### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by AmaliaGM**
## Project requirements

 n+1
- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. This is the corresponding wireframe view: blog all user posts
- Integration specs
    - Use Capybara to write integration tests for each view in your project.
    
    - User index page:
    I can see the username of all other users.
    I can see the profile picture for each user.
    I can see the number of posts each user has written.
    When I click on a user, I am redirected to that user's show page.

    - user show page:
    I can see the user's profile picture.
    I can see the user's username.
    I can see the number of posts the user has written.
    I can see the user's bio.
    I can see the user's first 3 posts.
    I can see a button that lets me view all of a user's posts.
    When I click a user's post, it redirects me to that post's show page.
    When I click to see all posts, it redirects me to the user's post's index page.
    
    - User post index page:
    I can see the user's profile picture.
    I can see the user's username.
    I can see the number of posts the user has written.
    I can see a post's title.
    I can see some of the post's body.
    I can see the first comments on a post.
    I can see how many comments a post has.
    I can see how many likes a post has.
    I can see a section for pagination if there are more posts than fit on the view.
    When I click on a post, it redirects me to that post's show page.
    
    - Post show page:
    I can see the post's title.
    I can see who wrote the post.
    I can see how many comments it has.
    I can see how many likes it has.
    I can see the post body.
    I can see the username of each commentor.
    I can see the comment each commentor left.

image

- RSPEC TESTING APROVE
image

- BEFORE FIXING
image

- AFTER FIXING 1rst ERROR
image

- AFTER FIXING 2nd ERROR
image

### CODE REVIEW

**Review by Davidosky007 — APPROVED** [review]
Hi @AmaliaGM @ErikStoupignan 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
* No linter errors 👍 
* Well written PR description 👍 
* Good looking readme file 👍 
* All tests passed ✔️ 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ class Post < ApplicationRecord
   validates :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
 
   def last_five_comments
-    comments.order(created_at: :desc).limit(5)
+    comments.includes(:author).order(created_at: :desc).limit(5)
```
Comment (Nemwel-Boniface) [inline app/models/post.rb:13]: - [ ] You have done a great job trying to resolve the N + 1 problem for fetching posts. However, to really make sure that this is fully resolved and meet all the project requirements **(Note:** N+1 problem is solved when fetching all posts and their comments for a user.) )you have to make this change as well in the index action in your app/controllers/posts_controller.rb file as this action is the one responsible for fetching all the posts.
- [ ] To customize it to fetch all the posts for a user including the comments as well (Now fixing the N + 1 problem) You can easily do it as below:

```
def index
    @user = User.find(params[:user_id])
    @userposts = @user.posts.includes(:comments)
  end
```

image

File: spec/integrations/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    it 'redirects to post show page when cliked' do
+      @user1 = User.create(name: 'Tom', photo: 'https://picsum.photos/seed/picsum/536/354', bio: 'Teacher from Mexico.')
+      @post1 = Post.create(title: 'First Post', text: 'This is my first post', likes_counter: 3, author: @user1)
+      visit "/users/#{@user1.id}/posts"
+      click_on Post
+      expect(page).to have_current_path(user_post_path(@user1.id, @post1.id))
+    end
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/integrations/user_spec.rb]: **For User Show:**

- [x] It appears that you may have forgotten to add specs for "I can see a button that lets me view all of a user's posts."

image

- [x] I can see you might have forgotten to add specs for "When I click to see all posts, it redirects me to the user's post's index page."

image

**Kindly include them.**

File: spec/integrations/post_spec.rb [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      visit(user_posts_path(@user1.id))
+      expect(page).to have_content('Comments: 3')
+    end
+
+    it 'can see how many likes a post has.' do
+      visit(user_posts_path(@user1.id))
+      expect(page).to have_content('Likes: 3')
+    end
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/integrations/post_spec.rb]: **For Post index**

- [x] I can see that you might have forgotten to add specs for "I can see a section for pagination if there are more posts than fit on the view."

image


- [x] - [ ] I can see that you might have forgotten to add specs for "When I click on a post, it redirects me to that post's show page"

image

**Kindly add them**

**Review by Nemwel-Boniface — CHANGES_REQUESTED** [review]
## Hello @ErikStoupignan     :wave: , 
I'm Nemwel your reviewer for this PR :man_beard: 

Good job working on your project so far! :clap: 
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights

Great job setting up the correct linters ✔️ 
Professional documentation ✔️ 
Correct workflow ✔️ 
Meaningful commit messages ✔️  


## ♻️ Required Changes ♻️

_Check the comments under the review._

## [Optional] suggestions

- [x] For some reason after following through the normal steps ie ```bundle rails db:create db:migrate``` and when I was going to now run the tests and see if they pass with ```rspec spec/integrations``` It was again requiring me to run ```bin/rails db:migrate RAILS_ENV=test```. This again threw another error which was coming from Bullet which is in config/environments/test.rb and disabled lines 10-12 and the tests now run. Is this an error that you also face?

image



_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding! :partying_face: 👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me **@Nemwel-Boniface** in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by AmaliaGM** [comment]
@Nemwel-Boniface this project was already approved. 41 min ago

**Comment by Davidosky007** [comment]
> @Nemwel-Boniface this project was already approved. 41 min ago

@AmaliaGM I missed some of the missing specs pointed out by @Nemwel-Boniface so I made a mistake 🙏 kindly fix the requested changes as you can see[ in the project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=Use%20Capybara%20to,each%20commentor%20left.) they are all required for this milestone

**Comment by AmaliaGM** [comment]
@Davidosky007 I will fix them. But already submitted the project.

File: app/models/post.rb (hunk 1 above)
Comment (ErikStoupignan) [reply app/models/post.rb:13]: @Nemwel-Boniface Thank you so much for your advice! We are already working on it.

Comment (ErikStoupignan) [reply app/models/post.rb:13]: I have finished applying your recommendation and the objective has become much clearer. Thank you very much!

File: spec/integrations/post_spec.rb (hunk 3 above)
Comment (ErikStoupignan) [reply spec/integrations/post_spec.rb]: Thank you very much for your observations. I have finished applying them.

File: spec/integrations/user_spec.rb (hunk 2 above)
Comment (ErikStoupignan) [reply spec/integrations/user_spec.rb]: Thank you very much for your observations. I have finished applying them.

**Comment by ErikStoupignan** [comment]
All recommendations were addressed and applied. Thank you very much for his advice.

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Team,  

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: . Congratulations!  🎉

Cheers..Happy coding! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**